Overhaul Component docs for v4 (#140)
HarperDB/documentation: docs/developers/components/managing.md

When developing a component locally there are a number of ways to run it on Harper.

### `dev` and `run` commands

The quickest way to run a component is by using the `dev` command within the component directory.

The `harperdb dev .` command will automatically watch for file changes within the component directory, and restart the Harper threads when changes are detected.

The `dev` command will **not** restart the main thread; if this is a requirement, switch to using `run` instead and manually start/stop the process to execute the main thread.

Stop execution for either of these processes by sending a SIGINT (generally CTRL/CMD+C) signal to the process.

### Deploying to a local Harper instance

Alternatively, to mimic interfacing with a hosted Harper instance, use operation commands instead.

1. Start up Harper with `harperdb`
2. _Deploy_ the component to the local instance by executing:
    ```sh
    harperdb deploy_component \
      project=<name> \
      package=<path-to-project> \
      restart=true \
      replicated=true
    ```
   - Make sure to omit the `target` option so that it _deploys_ to the Harper instance running locally

>    - The `package=<path>` option creates a symlink to the component simplifying restarts

Reviewer: We should probably make it clear that the default behavior of deploy_component is to deploy the current directory, and including `package` will deploy from that package reference (or is this trying to dictate a specific way of doing development?).

Author: I was particularly focussing on a simplified dev-experience here. Didn't really want to inundate with info about `deploy_component`, but I will review and consider how to improve this section

Author: Adding an additional info here - lmk your thoughts

>    - The `restart=true` option automatically restarts Harper threads after the component is deployed

Reviewer 2: `rolling` restarts might be a cool feature worth mentioning here

Author: What is that feature?

Reviewer 2: https://docs.harperdb.io/docs/developers/operations-api/components#deploy-component

for restart - "If set to rolling, a rolling restart will be triggered after the component is deployed"

Reviewer 3: Ah, the value can be a boolean or "rolling". Rad.

>    - The `replicated=true` option is important if running on multiple threads, but optional otherwise

Reviewer: That's correct Chris, `replicated` replicates the component to other nodes. This doesn't have anything to do with threads (which all should be restarted on a restart).

> 3. In another process, use the `harperdb restart` command to restart the instance's threads at any time.

Reviewer: What does "In another process" mean?

Author: Like another terminal - essentially I was going for "don't stop harper and run this command; instead open a new terminal and run this command"

Reviewer: If the intent is another terminal, maybe "In another terminal" would work?

Author: done

> Not all [component operations](../operations-api/components.md) are available via CLI. When in doubt, switch to using the Operations API via network requests to the local Harper instance.

For example, to properly _deploy_ a `test-component` locally, the command would look like:

```sh
harperdb deploy_component \
  project=test-component \
  package=/Users/dev/test-component \
  restart=true \
  replicated=true
```

> [!TIP]
> If the current directory is the component directory, use a shortcut such as `package=$(pwd)` to avoid typing out the complete path.

## Remote Management

Managing components on a remote Harper instance is best accomplished through [component operations](../operations-api/components.md), similar to using the `deploy_component` command locally. Before continuing, always backup critical Harper instances. Managing, deploying, and executing components can directly impact a live system.

Remote Harper instances work very similarly to local Harper instances. The primary component management operations still include `deploy_component`, `drop_component`, and `restart`.

> The key to remote management is using the `CLI_TARGET_USERNAME` and `CLI_TARGET_PASSWORD` environment variables, and the `target=<remote>` CLI option.

Reviewer: FWIW, you can  also use `username=<user>` and `password=<pass>` as command line arguments. I just prefer env variables, so that's what I have always suggested to people.

Author: I will include this as an option then!

The environment variables must be set to the remote instance's `username` and `password` respectively, and the `<remote>` must be set to the remote instance's URL.

All together:

```sh
export CLI_TARGET_USERNAME=<username>
export CLI_TARGET_PASSWORD=<password>
harperdb deploy_component \
  project=<name> \
  package=<package> # optional \
  target=<remote> \
  restart=true \
  replicated=true
```

Unlike local development where `package` should be set to a local file path for symlinking and improved development experience purposes, now it has some additional options.

A local component can be deployed to a remote instance by **omitting** the `package` field. Harper will automatically package the local directory and include that along with the rest of the deployment operation.

Furthermore, the `package` field can be set to any valid [npm dependency value](https://docs.npmjs.com/cli/v11/configuring-npm/package-json#dependencies).
  - For components deployed to npm, specify the package name: `package="@harperdb/status-check"`
  - For components on GitHub, specify the URL: `package="https://github.com/HarperDB/status-check"`, or the shorthand `package=HarperDB/status-check`

>   - Private repositories also work if the correct SSH keys are on the server: `package="git+ssh://git@github.com:HarperDB/secret-component.git"`

Reviewer: Probably include a link to the SSH keys operations.

> Before continuing, ensure Harper is installed and the `harperdb` CLI is available. For more information, review the [installation guide](../../deployments/install-harperdb/README.md).

When developing a component locally there are a number of ways to run it on Harper.

### `dev` and `run` commands

The quickest way to run a component is by using the `dev` command within the component directory.

The `harperdb dev .` command will automatically watch for file changes within the component directory and restart the Harper threads when changes are detected.

The `dev` command will **not** restart the main thread; if this is a requirement, switch to using `run` instead and manually start/stop the process to execute the main thread.

Stop execution for either of these processes by sending a SIGINT (generally CTRL/CMD+C) signal to the process.

### Deploying to a local Harper instance

Alternatively, to mimic interfacing with a hosted Harper instance, use operation commands instead.

1. Start up Harper with `harperdb`
2. _Deploy_ the component to the local instance by executing:
    ```sh
    harperdb deploy_component \

>       project={name} \

Reviewer 3: Not sure how I feel about the curly braces. Feels like a template. I'm so used to seeing `<` and `>`.

As I'm sure you know, other CLI tools use `<` and `>` to signifiy required args and `[` and `]` to signify optional args. I've never seen `{` and `}` used to describe CLI args.

# Managing Components

> Managing components on on Harper is easy. There are many ways to manage components with Harper, and some are limited to local development as compared to a hosted, managed Harper instance. This page will cover the recommended methods of developing, installing, deploying, and running Harper components.

Reviewer 3: I think the first sentence is unnecessary, but you can keep it if you remove the duplicate "on".

The second sentence reads weird. Consider rephrasing:

```suggestion
Harper offers several approaches to managing components that differ between local development and Harper managed instances. This page will cover the recommended methods of developing, installing, deploying, and running Harper components.
```

## Local Development

Harper is designed to be simple to run locally. Generally, Harper should be installed locally on a machine using a global package manager install (i.e. `npm i -g harperdb`).

> Before continuing, ensure Harper is installed and the `harperdb` CLI is available. For more information, review the [installation guide](../../deployments/install-harperdb/README.md).

When developing a component locally there are a number of ways to run it on Harper.

The quickest way to run a component is by using the `dev` or `run` commands (i.e. `harperdb run .`) within the component directory. Stop execution by sending a SIGINT (generally CTRL/CMD+C) signal to the process.

Alternatively, to mimic interfacing with a hosted Harper instance, use operation commands instead.

> 1. Start up Harper with `harperdb start`

Author: I'm removing `start` from this guide

2. "Deploy" the component to the local instance by executing `harperdb deploy_component project=<name>` from within the component directory
   1. Make sure to omit the `target` option so that it _deploys_ to the Harper instance running locally
3. Depending on the needs of the component
   1. If it needs a main thread restart, completely stop and start Harper again with `harperdb stop` and `harperdb start`
   2. If it only needs a worker thread restart, run `harperdb restart` (or include `restart=true` in the deploy command from step 2)

> 4. If changes need to be made to the component run `harperdb deploy_component project=<name>` again from within the component directory. It will automatically apply any changes.

Reviewer: Should we include the `replicated` and `restart` parameters here? (we use them a lot)

Reviewer 4: yes!

5. To remove the component use `harperdb drop_component project=<name>`

> Not all [component operations](../operations-api/components.md) are available via CLI. When in doubt, switch to using the Operations API via network requests to the local Harper instance.

## Hosted Harper Instance

> TODO

Reviewer: I assume `harperdb deploy_component` can be used here too.

## Advanced

The previous methodologies should be sufficient for most developers, but a more advanced process does exist. The following methods should be executed with caution as they can have unintended side-effects. Always backup any critical Harper instances before continuing.

> First, locate the Harper installation **rootPath** directory. Generally, this is `/hdb`. It can be retrieved by running `harperdb get_configuration` and looking for the **rootPath** field.

Reviewer 3: Is the path really `/hdb`? On my machine, it's `~/hdb`.

Author: fixed

> For a useful shortcut on UNIX machines run: `harperdb get_configuration json=true | jq ".rootPath" | sed 's/"//g'`

> This path is the Harper instance. Within this directory, locate the root config titled **harperdb-config.yaml**, and the components root path. The components root path will be `<rootPath>/components` by default, but it can also be configured. If necessary, use `harperdb get_configuration` again and look for the **componentsRoot** field for the exact path.

Reviewer 3: Bunch of nitpicks...

1. Instead of bolding text you want to stand out, consider using inline code backticks.
2. I don't think you need the comma after `harperdb-config.yaml`
3. Lead with "By default," (like you did with "If necessary") instead of in the middle of the sentence

```suggestion
This path is the Harper instance. Within this directory, locate the root config titled `harperdb-config.yaml` and the components root path. By default, the components root path will be `<rootPath>/components`, but it can also be configured. If necessary, use `harperdb get_configuration` again and look for the `componentsRoot` field for the exact path.
```

### Adding components to root

> Similar to how components can specify other components within their **config.yaml**, components can be added to Harper by adding them to the **harperdb-config.yaml**.

Reviewer 3: Inline code instead of bold?

```suggestion
Similar to how components can specify other components within their `config.yaml`, components can be added to Harper by adding them to the `harperdb-config.yaml`.
```

> The configuration is very similar to that of **config.yaml**. Entries are comprised of a top-level `<name>:`, and an indented `package: <specifier>` field. Any additional component option can also be included as indented fields.

Reviewer 3: Same with the bold. I don't think you need the comma after `<name>:`.

```suggestion
The configuration is very similar to that of `config.yaml`. Entries are comprised of a top-level `<name>:` and an indented `package: <specifier>` field. Any additional component option can also be included as indented fields.
```

```yaml
status-check:
  package: "@harperdb/status-check"
```

> The key difference between this and a component's **config.yaml** is that the name does **not** need to be associated with a **package.json** dependency. When Harper starts up, it transforms these configurations into a **package.json** file, and then executes a form of `npm install`. Thus, the `package: <specifier>` can be any valid dependency syntax such as npm packages, GitHub repos, tarballs, and local directories are all supported.

Reviewer 3: Inline code for `config.yaml` and `package.json`, but bold **not** is perfect. Removed unnecessary "and".

```suggestion
The key difference between this and a component's `config.yaml` is that the name does **not** need to be associated with a `package.json` dependency. When Harper starts up, it transforms these configurations into a `package.json` file, then executes a form of `npm install`. Thus, the `package: <specifier>` can be any valid dependency syntax such as npm packages, GitHub repos, tarballs, and local directories are all supported.
```

Given a root config like:

```yaml
myGithubComponent:
  package: HarperDB-Add-Ons/package#v2.2.0 # install from GitHub 
myNPMComponent:
  package: harperdb # install from npm
myTarBall:
  package: /Users/harper/cool-component.tar # install from tarball
myLocal:
  package: /Users/harper/local # install from local path
myWebsite:
  package: https://harperdb-component # install from URL
```

> Harper will generate a **package.json** like:

Reviewer 3: Inline code instead of bold?

```suggestion
Harper will generate a `package.json` like:
```

```json
{
  "dependencies": {
    "myGithubComponent": "github:HarperDB-Add-Ons/package#v2.2.0",
    "myNPMComponent": "npm:harperdb",
    "myTarBall": "file:/Users/harper/cool-component.tar",
    "myLocal": "file:/Users/harper/local",
    "myWebsite": "https://harperdb-component"
  }
}
```

> npm will install all the components and store them in `<rootPath>/<componentsRoot>`. A symlink back to `<rootPath>/node_modules` is also created for dependency resolution purposes.

Reviewer: technically `<componentsRoot>` = `<rootPath>/components` (by default) and `<rootPath>/<componentsRoot>` is just two full paths concatenated together.

The package prefix is automatically added, however you can manually set it in your package reference.

```yaml
myCoolComponent:
  package: file:/Users/harper/cool-component.tar
```

> By specifying a file path, Harper _should_ generate a symlink from that location to the **componentsRoot**. In case the symlink is not being generated, the final, advanced method for component loading is to manually create symlinks from the local component source directory to the Harper instance **componentsRoot**. For most UNIX machines this is as simple as: `ln -s <path/to/component/source> <componentsRoot>`.

Reviewer 3: I'm not sure I follow why the symlink wouldn't be created. Is that a bug?

The second sentence reads weird. I don't think you need the whole "final, advanced method" wording.

```suggestion
By specifying a file path, Harper _should_ generate a symlink from that location to the `componentsRoot`. In case the symlink is not being generated, manually create the symlink from the local component source directory to the Harper instance `componentsRoot*`. For most UNIX machines this is as simple as: `ln -s <path/to/component/source> <componentsRoot>`.
```

Author: You are right. I think I'm just going to remove this. We can consider npm not symlinking things as a bug.

> For example, if the machine had a **componentsRoot** of `/hdb/components`, and the component source was located at `/Users/dev/my-custom-component`, the symlink command would be: `ln -s /Users/dev/my-custom-component /hdb/components`

Reviewer 3: Inline comment instead of bold?

```suggestion
For example, if the machine had a `componentsRoot` of `/hdb/components`, and the component source was located at `/Users/dev/my-custom-component`, the symlink command would be: `ln -s /Users/dev/my-custom-component /hdb/components`
```
